internal/commands/document.go

```diff
@@ -50,97 +58,142 @@ Save the documentation to a specific directory
+		documentContents += "\n"
+	}
+
+	documentContents += "\n"
+
+	for _, document := range documentation {
+
+		// Title
+		documentContents += "## " + document.Header.Title
+		documentContents += "\n"
+
+		// Description
+		documentContents += document.Header.Description
+		documentContents += "\n"
+
+		// Resources
+		documentContents += "Resources: " + document.Header.Resources
+		documentContents += "\n\n"
+
+		// Rego
+		documentContents += "### Rego"
+		documentContents += "\n\n"
+
+		documentContents += "```rego"
+		documentContents += "\n"
+
+		documentContents += document.Rego
+		documentContents += "\n"
+
+		documentContents += "```"
+		documentContents += "\n"
+
+		// Source link
+		documentContents += "_source: [" + document.URL + "](" + document.URL + ")_"
+		documentContents += "\n\n"
+	}
+
+	documentContents = strings.TrimSuffix(documentContents, "\n")
+	if err := ioutil.WriteFile(viper.GetString("output"), []byte(documentContents), os.ModePerm); err != nil {
 		return fmt.Errorf("writing documentation: %w", err)
 	}
 
 	return nil
 }
 
-func getPolicyDocumentation(path string, outputDirectory string) (string, error) {
-	regoFilePaths, err := getRegoFilePaths(path)
-	if err != nil {
-		return "", fmt.Errorf("get rego files: %w", err)
-	}
-
-	policies, err := rego.LoadPolicies(regoFilePaths)
+func getDocumentation(path string, outputDirectory string) ([]Document, error) {
+	policies, err := rego.GetFilesWithAction(path, "violation")
 	if err != nil {
-		return "", fmt.Errorf("load policies: %w", err)
+		return nil, fmt.Errorf("get files: %w", err)
 	}
 
-	policyDocument := "# Policies\n\n"
-	policyDocument += "|Name|Rule Types|API Groups|Kinds|Description|\n"
-	policyDocument += "|---|---|---|---|---|\n"
-
+	var documents []Document
 	for _, policy := range policies {
-		policyCommentBlocks, err := getPolicyCommentBlocks(policy.Comments)
+		header, err := getHeader(policy.Comments)
 		if err != nil {
-			return "", fmt.Errorf("get policy comment blocks: %w", err)
+			return nil, fmt.Errorf("get header: %w", err)
 		}
 
-		for _, policyCommentBlock := range policyCommentBlocks {
-			relPath, err := filepath.Rel(outputDirectory, policy.FilePath)
-			if err != nil {
-				return "", fmt.Errorf("rel path: %w", err)
-			}
+		relPath, err := filepath.Rel(outputDirectory, policy.FilePath)
+		if err != nil {
+			return nil, fmt.Errorf("rel path: %w", err)
+		}
+		relDir := filepath.Dir(relPath)
 
-			relDir := filepath.Dir(relPath)
-			ruleTypes := strings.Join(policy.RulesActions, ", ")
-			apiGroups := strings.Join(policyCommentBlock.APIGroups, ", ")
-			kinds := strings.Join(policyCommentBlock.Kinds, ", ")
-
-			policyDocument += fmt.Sprintf("|[%s](%s)|%s|%s|%s|%s|\n",
-				getNameFromPath(policy.FilePath),
-				relDir,
-				ruleTypes,
-				apiGroups,
-				kinds,
-				policyCommentBlock.Description,
-			)
+		regoWithoutComments := getRegoWithoutComments(policy.Contents)
```

Comment:
I originally removed them for the documentation, because I didn't want the header to be repeated. Then I stumbled upon another win, which was removing them when embedding them into the constraints.

I can't say I'm too strongly opinionated one way or the other, I agree that at a minimum we need removeHeader(). I don't know if we need both removeHeader _and_ removeAll(). Or if we just leave inline comments in the constraints.

Reply:
I would prefer the comments are left in the constraints. If there's something important enough to write a comment for, I would want to see it in the Kubernetes resource if I were troubleshooting and did not write the policy.

Reply:
Going to table this for now as I don't think this should block the PR, we can discuss further later.